File: 1er Parcial/Practica1.py
# Problema del Puente y la Antorcha
# Estrategia óptima: siempre usar los dos más rápidos como "mensajeros"
# y en cada paso decidir la opción que minimiza el tiempo.

def cruzar_puente(tiempos):
    # Ordenamos los tiempos de menor a mayor
    tiempos.sort()
    pasos = []   # Aquí guardamos los movimientos
    total = 0    # Tiempo total

    # Mientras haya más de 3 personas en la orilla inicial
    while len(tiempos) > 3:
        a = tiempos[0]         # El más rápido
        b = tiempos[1]         # El segundo más rápido
        x = tiempos[-2]        # El penúltimo más lento
        y = tiempos[-1]        # El más lento

        # Estrategia 1: a y b cruzan, a regresa, x y y cruzan, b regresa
        tiempo_op1 = b + a + y + b

        # Estrategia 2: a y y cruzan, a regresa, a y x cruzan, a regresa
        tiempo_op2 = y + a + x + a

        # Elegimos la mejor estrategia
        if tiempo_op1 <= tiempo_op2:
            # Estrategia 1
            pasos.append(f"({a},{b}) -> {b}")
            pasos.append(f"{a} <- {a}")
            pasos.append(f"({x},{y}) -> {y}")
            pasos.append(f"{b} <- {b}")
            total += tiempo_op1
        else:
            # Estrategia 2
            pasos.append(f"({a},{y}) -> {y}")
            pasos.append(f"{a} <- {a}")
            pasos.append(f"({a},{x}) -> {x}")
            pasos.append(f"{a} <- {a}")
            total += tiempo_op2

        # Sacamos a los dos más lentos (ya cruzaron)
        tiempos = tiempos[:-2]

    # Casos finales (<=3 personas)
    if len(tiempos) == 3:
        a, b, c = tiempos
        pasos.append(f"({a},{b}) -> {b}")
        pasos.append(f"{a} <- {a}")
        pasos.append(f"({a},{c}) -> {c}")
        total += a + b + c
    elif len(tiempos) == 2:
        a, b = tiempos
        pasos.append(f"({a},{b}) -> {b}")
        total += b
    elif len(tiempos) == 1:
        pasos.append(f"{tiempos[0]} -> {tiempos[0]}")
        total += tiempos[0]

    return pasos, total


# ------------------- PROGRAMA PRINCIPAL -------------------
if __name__ == "__main__":
    n = int(input("Número de personas: "))
    tiempos = []
    print("Ingresa los tiempos de cruce:")
    for _ in range(n):
        tiempos.append(int(input()))

    pasos, total = cruzar_puente(tiempos)

    # Mostrar el recorrido
    print("\nRecorrido óptimo:")
    for p in pasos:
        print(p, end="; ")

    print(f"Total = {total}")

File: 1er Parcial/Practica1.c
#include <stdio.h>
#include <stdlib.h>

int comparar(const void *a, const void *b) {
    return (*(int*)a - *(int*)b);
}

int main() {
    int n, i;
    printf("Numero de personas: ");
    scanf("%d", &n);

    int tiempos[n];
    printf("Ingresa los tiempos de cruce:\n");
    for(i = 0; i < n; i++)
        scanf("%d", &tiempos[i]);

    // Ordenar tiempos de menor a mayor
    qsort(tiempos, n, sizeof(int), comparar);

    int total = 0;
    int izq = n - 1;
    int primero = 1; // para manejar formato de "; "

    while(izq >= 3) {
        int a = tiempos[0];     // más rápido
        int b = tiempos[1];     // segundo más rápido
        int x = tiempos[izq-1]; // penúltimo más lento
        int y = tiempos[izq];   // más lento

        int opcion1 = 2*b + a + y;
        int opcion2 = 2*a + x + y;

        if(opcion1 <= opcion2) {
            if(!primero) printf("; ");
            printf("(%d,%d) -> %d; %d <- %d; (%d,%d) -> %d; %d <- %d", 
                   a, b, b, a, a, x, y, y, b, b);
            total += opcion1;
            primero = 0;
        } else {
            if(!primero) printf("; ");
            printf("(%d,%d) -> %d; %d <- %d; (%d,%d) -> %d; %d <- %d",
                   a, y, y, a, a, a, x, x, a, a);
            total += opcion2;
            primero = 0;
        }
        izq -= 2;
    }

    if(izq == 2) {
        if(!primero) printf("; ");
        printf("(%d,%d) -> %d; %d <- %d; (%d,%d) -> %d",
               tiempos[0], tiempos[1], tiempos[1],
               tiempos[0], tiempos[0],
               tiempos[0], tiempos[2], tiempos[2]);
        total += tiempos[0] + tiempos[1] + tiempos[2];
    } else if(izq == 1) {
        if(!primero) printf("; ");
        printf("(%d,%d) -> %d", tiempos[0], tiempos[1], tiempos[1]);
        total += tiempos[1];
    } else if(izq == 0) {
        if(!primero) printf("; ");
        printf("(%d) -> %d", tiempos[0], tiempos[0]);
        total += tiempos[0];
    }

    printf(". Total = %d\n", total);

    return 0;
}
